### PR TITLE
Grok 1

### PR DESCRIPTION
**Pull request by yieldthought**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/9442)

### Problem description
Experimental implementation for Grok-1 314B model bringup on T3000

### What's changed
Modified version of the Mixtral implementation for the various Grok changes. Still in the experimental folder until performance passes have been made and demo brought up on Galaxy, but for now correctness seems good and it's time to merge to main to support other refactoring efforts

### Checklist
- [x] Post commit CI passes
- [x] [Model regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/10212684366)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

File: models/experimental/grok/tt/grok_attention.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        x_11BH = xs
+        wo = self.wo
+        layer_past = self.layer_past
+        rot_mat = rot_mats[current_pos]
+        attn_mask_1B4P = attn_masks
+        ###
+        # QKV matmuls
+        ###
+
+        xqkv_fused = ttnn.experimental.operations.primary.matmul(
```
Comment (ayerofieiev-tt) [inline models/experimental/grok/tt/grok_attention.py:171]: this can be just `ttnn.matmul`

**Review by ayerofieiev-tt — COMMENTED** [review]
This is really nice! 

One painful observation 😫 Torch/TTNN entries: 259 / 273

File: models/experimental/grok/tt/grok_attention.py (hunk 1 above)
Comment (yieldthought) [reply models/experimental/grok/tt/grok_attention.py]: Thanks, rebased and updated all of these!

**Comment by yieldthought** [comment]
Added unit tests, check this goes green:
- [ ] https://github.com/tenstorrent/tt-metal/actions/runs/9971962402

**Comment by ttmchiou** [comment]
looks like https://github.com/tenstorrent/tt-metal/actions/runs/9971962402 failed?

File: .github/workflows/t3000-unit-tests.yaml [hunk 2]
```diff
@@ -22,6 +22,7 @@ jobs:
           { name: "t3k falcon7b tests", arch: wormhole_b0, cmd: run_t3000_falcon7b_tests, timeout: 30, owner_id: UBHPP2NDP}, #Joseph Chu
           { name: "t3k falcon40b tests", arch: wormhole_b0, cmd: run_t3000_falcon40b_tests, timeout: 30, owner_id: U044T8U8DEF}, #Johanna Rock
           { name: "t3k mixtral tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 30, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
+          { name: "t3k grok tests", arch: wormhole_b0, cmd: run_t3000_grok_tests, timeout: 30, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
```
Comment (ttmchiou) [inline .github/workflows/t3000-unit-tests.yaml:25]: will need a slack owner here. Are you the owner @yieldthought (your slack member ID is U03HY7MK4BT)
Owners are pinged using their own slack ID and are pinged in a slack chat if their test fails on main branch.
https://forums.slackcommunity.com/s/question/0D5Hq00009mZfnuKAC/hello-i-am-a-slack-newbie-how-do-i-find-my-slack-id?language=en_US

File: models/experimental/grok/demo/demo.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+import os
+import torch
+import json
+import tt_lib as ttl
+import pytest
+from loguru import logger
+from time import time
+
+# Set Grok flags for CI, if CI environment is setup
+if os.getenv("CI") == "true":
```
Comment (ttmchiou) [inline models/experimental/grok/demo/demo.py:13]: https://github.com/tenstorrent/tt-metal/pull/10371/files
we have a utility function recently merged for checking if "CI")

Comment (yieldthought) [reply models/experimental/grok/demo/demo.py:13]: That function cannot be used in our code. The purpose of these checks are to set environment variables that are necessary before ttnn is imported. However, conftest.py has a complex import chain that ends up importing ttnn to implement is_grayskull. Even if that were resolved, relying on nobody in any of those upstream modules adding a ttnn import in the future would be... wildly hopeful.

File: .github/workflows/t3000-unit-tests.yaml (hunk 2 above)
Comment (yieldthought) [reply .github/workflows/t3000-unit-tests.yaml]: Added my slack ID, thanks.

File: tests/scripts/t3000/run_t3000_unit_tests.sh [hunk 4]
```diff
@@ -112,21 +112,45 @@ run_t3000_mixtral_tests() {
...
-  run_t3000_falcon7b_tests
+#   # Run falcon40b tests
+#   run_t3000_falcon40b_tests
 
-  # Run falcon40b tests
-  run_t3000_falcon40b_tests
+#   # Run mixtral tests
+#   run_t3000_mixtral_tests
 
-  # Run mixtral tests
```
Comment (ttmchiou) [inline tests/scripts/t3000/run_t3000_unit_tests.sh:128]: I don't think these tests are suppose to be removed?

File: models/experimental/grok/tests/test_grok_decoder.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+import ttnn
+from models.experimental.grok.tt.grok_common import prepare_inputs_ttnn, prepare_rotation_mat_ttnn
+from models.experimental.grok.tt.grok_decoder import TtTransformerBlock
+from models.experimental.grok.reference.model import DecoderLayer
+from models.experimental.grok.tt.model_config import TtModelArgs
+from models.utility_functions import comp_pcc, comp_allclose
+from ttnn import ReplicateTensorToMesh, ConcatMeshToTensor
+
+
+@pytest.mark.timeout(500 * 8)
```
Comment (ttmchiou) [inline models/experimental/grok/tests/test_grok_decoder.py:26]: Can we remove these timeouts?
We use xdist for timeouts which handles inadvertent hangs/crashes for pytests.

Follows the format such as this ripped from run_t3000_demo_tests.sh

Comment (yieldthought) [reply models/experimental/grok/tests/test_grok_decoder.py:26]: Sure I can move them to the script level

**Review by ttmchiou — APPROVED** [review]
Pre-emptively approving to not block this.
take a look at pytest.mark.timeout if that can be looked at

**Pull request merged**